python/ray/util/sgd/data/dataset.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+
+import random
+import ray
+import torch
+
+from ray.util.sgd.data.utils import get_arr_partition
+
+
+@ray.remote
+def _load_remote(dataset, path):
+    return dataset.load(path)
+
+
+class Dataset(ABC, torch.utils.data.IterableDataset):
+    """
+    This class defines the Ray dataset abstration (primarily for SGD).
+    It's primarily meant to be treated as an abstract class.
+    """
+
+    def __init__(self, loc, remote=False, max_paths=None, transform=None):
+        """Needs to be initialized with a dataset location
+
+        Args:
+            loc: Location of training data (i.e. root of training directory)
+                This is passed into get_paths
+            remote: Whether dataset loading
+                should be parallelized on multiple workers/nodes. For example,
+                if only the driver has the necessary permissions to access
+                the data, remote should be False
+            max_paths: Maximum number of paths to load (passed to `get_paths`)
+
+        """
+        self._paths = self.get_paths(loc, max_paths=max_paths)
+        self.transform = transform
+
+        if remote:
+            data = [_load_remote.remote(self, path) for path in self._paths]
+        else:
+            data = [ray.put(self.load(path)) for path in self._paths]
+
+        random.seed(0)
+        random.shuffle(data)
```

Comment:
Making a note here: we'd eventually want to use Parallel Iterator's shuffling here.